src/Processors/Transforms/FilterTransform.cpp

```diff
@@ -34,6 +36,141 @@ static void replaceFilterToConstant(Block & block, const String & filter_column_
+            while (end - begin >= 32)
+            {
+                _mm256_storeu_si256(
+                    reinterpret_cast<__m256i *>(begin),
+                    _mm256_and_si256(
+                        _mm256_loadu_si256(reinterpret_cast<const __m256i *>(begin)),
+                        _mm256_loadu_si256(reinterpret_cast<const __m256i *>(selected_by_filter))));
+                description->has_one |= !memoryIsZero(begin, 0, 32);
+                begin += 32;
+                selected_by_filter += 32;
+            }
+#elif defined(__SSE2__)
+            while (end - begin >= 16)
+            {
+                _mm_storeu_si128(
+                    reinterpret_cast<__m128i *>(begin),
+                    _mm_and_si128(
+                        _mm_loadu_si128(reinterpret_cast<const __m128i *>(begin)),
+                        _mm_loadu_si128(reinterpret_cast<const __m128i *>(selected_by_filter))));
+                description->has_one |= !memoryIsZero(begin, 0, 16);
+                begin += 16;
+                selected_by_filter += 16;
+            }
+#endif
+
+            while (end - begin >= 8)
+            {
+                *reinterpret_cast<UInt64 *>(begin) &= *reinterpret_cast<const UInt64 *>(selected_by_filter);
+                description->has_one |= *reinterpret_cast<UInt64 *>(begin);
+                begin += 8;
+                selected_by_filter += 8;
+            }
+
+            while (end - begin > 0)
+            {
+                *begin &= *selected_by_filter;
+                description->has_one |= *begin;
+                begin++;
+                selected_by_filter++;
+            }
+
+            description->data_holder = std::move(mutable_holder);
+            description->data = &data;
+        }
+    }
+    return std::move(description);
+}
+
+static std::unique_ptr<IFilterDescription> combineFilterAndIndices(
+    std::unique_ptr<SparseFilterDescription> description,
+    std::shared_ptr<const ChunkSelectFinalIndices> & select_final_indices_info,
+    size_t num_rows)
+{
+    /// Iterator interface to decorate data from output of std::set_intersection
+    struct Iterator
+    {
+        UInt8 * data;
+        Int64 & pop_cnt;
+        explicit Iterator(UInt8 * data_, Int64 & pop_cnt_) : data(data_), pop_cnt(pop_cnt_) {}
+        Iterator & operator = (UInt64 index) { data[index] = 1; ++pop_cnt; return *this; }
+        Iterator & operator ++ () { return *this; }
+        Iterator & operator * () { return *this; }
+    };
+
+    if (select_final_indices_info)
+    {
+        const auto * index_column = select_final_indices_info->select_final_indices;
+
+        if (index_column && index_column->size() != num_rows && description->hasOne())
+        {
+            std::unique_ptr<FilterDescription> res;
+            res->has_one = 0;
+            const auto & selected_by_indices = index_column->getData();
+            const auto & selected_by_filter = description->filter_indices->getData();
```

Comment:
should we check that `description` and `description->filter_indices` are not-null?

Reply:
This has been checked in `hasOne`